nameserver/addrs.go

```diff
@@ -0,0 +1,99 @@
+package nameserver
+
+import (
+	"bytes"
+	"fmt"
+	. "github.com/weaveworks/weave/common"
+	"net"
+)
+
+// A zone record
+// Note: we will try to keep all the zone records information is concentrated here. Maybe not all queries will
+//       use things like priorities or weights, but we do not want to create a hierarchy for dealing with all
+//       possible queries...
+type ZoneRecord interface {
+	Name() string  // The name for this IP
+	IP() net.IP    // The IP (v4) address
+	Priority() int // The priority
+	Weight() int   // The weight
+}
+
+type ZoneLookup interface {
+	// Lookup for a name
+	LookupName(name string) ([]ZoneRecord, error)
+	// Lookup for an address
+	LookupInaddr(inaddr string) ([]ZoneRecord, error)
+}
+
+type ZoneObserver interface {
+	// Observe anything that affects a particular name in the zone
+	ObserveName(name string, observer ZoneRecordObserver) error
+	// Observe anything that affects a particular IP in the zone
+	ObserveInaddr(inaddr string, observer ZoneRecordObserver) error
+}
+
+type ZoneRecordObserver func(ZoneRecord)
+
+/////////////////////////////////////////////////////////////
+
+// A basic record struct (satisfying the ZoneRecord interface)
+type Record struct {
+	name     string
+	ip       net.IP
+	priority int
+	weight   int
+}
+
+func (r Record) Name() string  { return r.name }
+func (r Record) IP() net.IP    { return r.ip }
+func (r Record) Priority() int { return r.priority }
+func (r Record) Weight() int   { return r.weight }
+
+func (i Record) String() string {
+	var buf bytes.Buffer
+	if len(i.Name()) > 0 {
+		buf.WriteString(fmt.Sprintf("%s", i.Name()))
+	}
+	if !i.IP().IsUnspecified() {
+		buf.WriteString(fmt.Sprintf("[%s]", i.IP()))
+	}
+	if i.Priority() > 0 {
+		buf.WriteString(fmt.Sprintf("/P:%d", i.Priority()))
+	}
+	if i.Weight() > 0 {
+		buf.WriteString(fmt.Sprintf("/W:%d", i.Weight()))
+	}
+
+	return buf.String()
+}
+
+/////////////////////////////////////////////////////////////
+
+// simple IPv4 type that can be used as a key in a map (in contrast with net.IP), used for sets of IPs, etc...
```

Comment:
This type, `IPv4`, will not be really used until I introduce the new `ZoneDb` internal database format...